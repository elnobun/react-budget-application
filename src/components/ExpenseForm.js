import React, { useState } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

const ExpenseForm = (props) => {
  // State
  const [description, setDescription] = useState(
    props.expense ? props.expense.description : ""
  );
  const [note, setNote] = useState(props.expense ? props.expense.note : "");
  const [amount, setAmount] = useState(
    props.expense ? props.expense.amount / 100 : ""
  );
  const [createdAt, setCreatedAt] = useState(
    props.expense ? moment(props.expense.createdAt) : moment()
  );
  const [calenderFocused, setCalenderFocused] = useState(false);
  const [error, setError] = useState("");

  // Functions
  const handleChangeDescription = (e) => {
    const description = e.target.value;
    setDescription(description);
  };

  const handleNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  };

  const handleAmountChange = (e) => {
    const amount = e.target.value;
    !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ? setAmount(amount) : null; // regex101.com
  };

  const handleDateChange = (date) => {
    date ? setCreatedAt(date) : null;
  };

  const handleFocusChange = ({ focused }) => {
    setCalenderFocused(focused);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!description || !amount) {
      setError("Please provide description and amount.");
    } else {
      setError("");
      props.onSubmit({
        description,
        note,
        amount: parseFloat(amount) * 100,
        createdAt: createdAt.valueOf(),
      });
    }
  };

  return (
    <form className="form" onSubmit={handleSubmit}>
      {error && <p className="form__error">{error}</p>}
      <input
        type="text"
        className="text-input"
        placeholder="Description"
        value={description}
        onChange={handleChangeDescription}
        autoFocus
      />
      <input
        type="text"
        className="text-input"
        placeholder="Amount"
        value={amount}
        onChange={handleAmountChange}
      />
      <SingleDatePicker
        date={createdAt}
        onDateChange={handleDateChange}
        focused={calenderFocused}
        onFocusChange={handleFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
      <textarea
        className="textarea"
        placeholder="Add a note for your expense (optional) "
        value={note}
        onChange={handleNoteChange}
      />
      <div>
        <button className="button">
          {props.expense ? "Edit Expense" : "Add Expense"}
        </button>
      </div>
    </form>
  );
};

export default ExpenseForm;
