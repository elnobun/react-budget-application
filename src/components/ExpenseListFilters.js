import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  setEndDate,
  setStartDate,
  setTextFilter,
  sortByAmount,
  sortByDate,
} from "../redux/actions/filters";
import { DateRangePicker, SingleDatePicker } from "react-dates";
import "react-dates/initialize";

const ExpenseListFilters = () => {
  // State
  const [calenderFocused, setCalenderFocus] = useState(null);

  // Redux Store
  const filters = useSelector((state) => state.filters);
  const dispatch = useDispatch();

  // Functions
  const handleInputChange = (e) => {
    const userInput = e.target.value;
    dispatch(setTextFilter(userInput));
  };

  const handleSelectChange = (e) => {
    const userSelect = e.target.value;
    userSelect === "date"
      ? dispatch(sortByDate())
      : userSelect === "amount"
      ? dispatch(sortByAmount())
      : null;
  };

  const handleDateChange = ({ startDate, endDate }) => {
    dispatch(setStartDate(startDate));
    dispatch(setEndDate(endDate));
  };

  const handleFocusChange = (calenderFocused) => {
    setCalenderFocus(calenderFocused);
  };

  return (
    <div>
      <input
        type="text"
        placeholder="Filter Expenses"
        value={filters.text}
        onChange={handleInputChange}
      />

      <select value={filters.sortBy} onChange={handleSelectChange}>
        <option value="" hidden>
          Sort Expenses
        </option>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>

      <DateRangePicker
        startDate={filters.startDate}
        startDateId="start"
        endDate={filters.endDate}
        endDateId="end"
        onDatesChange={handleDateChange}
        focusedInput={calenderFocused}
        onFocusChange={handleFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </div>
  );
};

export default ExpenseListFilters;
