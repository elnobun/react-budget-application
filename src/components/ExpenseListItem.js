import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <h3>
        <Link to={`edit/${id}`}>{description}</Link>
      </h3>
      <p>
        amount: {amount} - createdAt: {createdAt}
      </p>
    </div>
  );
};

export default ExpenseListItem;
