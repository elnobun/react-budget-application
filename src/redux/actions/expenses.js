import { v4 as uuid } from "uuid";

// ADD EXPENSE
export const addExpense = (payload) => ({
  type: "ADD_EXPENSE",
  expense: {
    id: uuid(),
    description: payload.description || "",
    note: payload.note || "",
    amount: payload.amount || 0,
    createdAt: payload.createdAt || 0,
  },
});

// REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

// EDIT EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update,
});
