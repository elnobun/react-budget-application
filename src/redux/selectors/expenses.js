// Get the visible expense
import moment from "moment";

const selectExpenses = (expenses, filters) => {
  // filter the expense array
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      // Match start & end date
      const startDateMatch = filters.startDate
        ? filters.startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = filters.endDate
        ? filters.endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      // Match expenses description and filters text input
      const textMatch = expense.description
        .toLowerCase()
        .includes(filters.text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (filters.sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1; // descending order - lower 1st
      } else if (filters.sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1; //descending order
      }
    });
};

export default selectExpenses;
