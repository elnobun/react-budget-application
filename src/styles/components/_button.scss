.button {
  border: none;
  border-radius: 0;
  background-color: $google-red;
  color: white;
  display: inline-block;
  font-size: $font-size-large;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary {
  background-color: #888888;
}
