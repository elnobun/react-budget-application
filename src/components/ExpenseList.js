import React from "react";
import { useSelector } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../redux/selectors/expenses";

export const ExpenseList = () => {
  const expenses = useSelector((state) =>
    selectExpenses(state.expenses, state.filters)
  );
  return (
    <div>
      <h1>Expense List</h1>
      {expenses.length === 0 ? (
        <p>No Expenses</p>
      ) : (
        expenses.map((expense) => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  );
};

export default ExpenseList;
