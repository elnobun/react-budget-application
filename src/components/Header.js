import React from "react";
import { Link } from "react-router-dom";
import { useDispatch } from "react-redux";
import { startLogout } from "../redux/actions/auth";

const Header = () => {
  const dispatch = useDispatch();

  return (
    <header className="header">
      <div className="container">
        <div className="header__content">
          <Link className="header__title" to="/dashboard">
            <h1>
              <i className="fad fa-coins" /> Budget Manager
            </h1>
          </Link>
          <button
            className="button button--link"
            onClick={dispatch(startLogout)}
          >
            <i className="fad fa-sign-out-alt" /> Logout
          </button>
        </div>
      </div>
    </header>
  );
};

export default Header;
