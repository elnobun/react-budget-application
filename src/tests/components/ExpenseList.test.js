import React from "react";
import { ExpenseList } from "../../components/ExpenseList";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { ReduxProvider } from "../setupProvider";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(
    <ReduxProvider>
      <ExpenseList expenses={expenses} />
    </ReduxProvider>
  );
  expect(wrapper).toMatchSnapshot();
});
