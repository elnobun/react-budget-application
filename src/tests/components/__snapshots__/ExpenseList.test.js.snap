// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<Provider>
  <ExpenseList
    expenses={
      Array [
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Rum",
          "id": 1,
          "note": "",
        },
        Object {
          "amount": 109500,
          "createdAt": -345600000,
          "description": "Coffee",
          "id": 2,
          "note": "",
        },
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "Credit card",
          "id": 3,
          "note": "",
        },
      ]
    }
  />
</Provider>
`;
